File: sherry/week-4/615. Course Schedule.java
// 615. Course Schedule
//
// There are a total of n courses you have to take, labeled from 0 to n - 1.
//
// Some courses may have prerequisites, for example to take course 0 you have to first take course 1, which is expressed as a pair: [0,1]
//
// Given the total number of courses and a list of prerequisite pairs, is it possible for you to finish all courses?
//
// Example
//
// Given n = 2, prerequisites = [[1,0]]
// Return true
//
// Given n = 2, prerequisites = [[1,0],[0,1]]

//TODO: test version, using binary search to increase time complexity
public class Solution {
    /*
     * @param numCourses: a total of n courses
     * @param prerequisites: a list of prerequisite pairs
     * @return: true if can finish all courses or false
     */
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // write your code here
        //prerequesities are direct nodes. so
        // 1. loop over to find all indegree, put in map<int, indegree>
        // 2. initailize a dequeue to search sort
        // 3. degree the ingree ->
        // return true if dequeue is null

        if(prerequisites == null || prerequisites.length == 0) {
            return true;
        }

        HashMap<Integer, Integer> mapping = new HashMap();

        for(int[] pre : prerequisites) {
            //if [1, 0] -> incress degree of 0;
            if(mapping.containsKey(pre[1])){
                mapping.put(pre[1], mapping.get(pre[1])+1);
            }else{
                mapping.put(pre[1], 1);
            }
        }
        Arrays.sort(prerequisites, arrayComparator);
        Queue<Integer> dequeue = new LinkedList();
        for(int i = 0; i < numCourses; i++) {
            if(!mapping.containsKey(i)){
                dequeue.offer(i);
            }
        }
        // for(int i = 0; i < prerequisites.length; i++) {
        //     System.out.println(prerequisites[i][0] + " " + prerequisites[i] [1]);
        // }
        // how to deal with duplicate code
        // ERROR for duplicate code
        while(!dequeue.isEmpty()) {
            int head = dequeue.poll();
            // useing binary search to get
            if(getArrayForIndex(head, prerequisites) != null){
                for(int[] pre: getArrayForIndex(head, prerequisites)){
                    if(pre[0] == head) {
                        mapping.put(pre[1], mapping.get(pre[1]) - 1);
                        if(mapping.get(pre[1]) == 0) {
                            dequeue.offer(pre[1]);
                            mapping.remove(pre[1]);
                        }
                    }
                }
            }
        }
        if(mapping.isEmpty()) {
            return true;
        }
        return false;
    }

    final Comparator<int[]> arrayComparator = new Comparator<int[]>() {
        @Override
        public int compare(int[] o1, int[] o2) {
            return o1[0]-o2[0];
        }
    };

    private int[][] getArrayForIndex(int target, int[][] prerequisites) {
        int[][] result;
        int start = 0;
        int end = prerequisites.length - 1;
        int left = 0;
        int right = end;
        while(start + 1 < end) {
            int mid = start + (end - start) / 2;
            if(prerequisites[mid][0] >= target) {
                end = mid;
            }else{
                start = mid;
            }
        }
        if(prerequisites[start][0] == target) {
            left = start;
        }else {
            left = end;
        }
        start = 0;
        end = prerequisites.length - 1;
         while(start + 1 < end) {
            int mid = start + (end - start) / 2;
            if(prerequisites[mid][0] <= target) {
                start = mid;
            }else{
                end = mid;
            }
        }
        if(prerequisites[end][0] == target) {
            right = end;
        }else {
            right = start;
        }
        if(right - left + 1 > 0) {
        result = new int[right-left + 1][2];
            for(int i = 0; i < result.length; i++) {
                result[i] = prerequisites[left + i].clone();
            }
            return result;
        }
        return null;
    }

}

File: Joe Zheng/week 4/615. Course Schedule.py
from collections import deque


class Solution:
    """
    @param: numCourses: a total of n courses
    @param: prerequisites: a list of prerequisite pairs
    @return: true if can finish all courses or false
    """

    def canFinish(self, numCourses, prerequisites):
        # write your code here。
        if numCourses == None or prerequisites == None or numCourses <= 0:
            return True

        # 统计所有点的入度，并初始化拓扑序列为空。
        edges = {i: [] for i in range(numCourses)}
        indegree = [0 for _ in range(numCourses)]

        # 每一段preqeruisite只是2个，是一个pair
        # 在有向图中，如果存在一条有向边 A-->B，那么我们认为这条边为 A 增加了
        # 一个出度，为 B 增加了一个入度
        # 先完成prere再完成course，所有是 preRe --> course
        for course, preRe in prerequisites:
            edges[preRe].append(course)
            indegree[course] += 1

        queue = deque()

        # 将所有入度为 0 的点，也就是那些没有任何依赖的点，
        # 放到宽度优先搜索的队列中
        for n in range(numCourses):
            # no dependent courses
            if indegree[n] == 0:
                queue.append(n)

        count = 0
        # 将队列中的点一个一个的释放出来，放到拓扑序列中，每次释放出某个点 A
        # 的时候，就访问 A 的相邻点（所有A指向的点），并把这些点的入度减去 1。
        while queue:
            course = queue.popleft()
            count += 1
            for edge in edges[course]:
                indegree[edge] -= 1
                # 如果发现某个点的入度被减去 1 之后变成了 0，则放入队列中。
                if indegree[edge] == 0:
                    queue.append(edge)
        # 直到队列为空时，算法结束，
        return count == numCourses

# 2
# [[1,0]]
